### PR TITLE
[DO-1660]Splitted Sanction model

### PR DESCRIPTION
**Pull request by IegorSoboliev**

### CODE REVIEW

File: business_register/models/sanction_models.py [hunk 1]
```diff
@@ -28,84 +28,97 @@ class Meta:
...
+        verbose_name = _('Sanction against country')
+        verbose_name_plural = _('Sanctions against countries')
+        ordering = ['start_date']
+
+
+class PersonSanction(Sanction):
+    is_foreign = models.BooleanField(
+        _('is foreign'),
+        db_index=True,
+        help_text=_('is foreign or Ukrainian')
```
Comment (criticus) [inline business_register/models/sanction_models.py:74]: consider to add a default value

File: business_register/models/sanction_models.py [hunk 2]
```diff
@@ -28,84 +28,97 @@ class Meta:
...
+    )
+    pep = models.ForeignKey(
```
Comment (criticus) [inline business_register/models/sanction_models.py:76]: are you adding this as a two-side connection? Sanctions under PEP, and PEP under sanctions, correct?

File: business_register/models/sanction_models.py [hunk 3]
```diff
@@ -28,84 +28,97 @@ class Meta:
...
     )
-    object_name = models.CharField(
-        _('object name'),
-        max_length=100,
-        db_index=True,
-        help_text=_('name of of the object under sanctions')
+    first_name = models.CharField(
+        _('first name'),
+        max_length=20,
+        help_text='first name of of the person under sanctions'
```
Comment (criticus) [inline business_register/models/sanction_models.py:94]: typo

File: business_register/models/sanction_models.py [hunk 4]
```diff
@@ -28,84 +28,97 @@ class Meta:
...
-        max_length=100,
-        db_index=True,
-        null=True,
-        blank=True,
-        default=None,
-        help_text=_('name of the object under sanctions in the language of the country of origin')
+    middle_name = models.CharField(
+        _('middle name'),
+        max_length=25,
+        help_text='middle name of of the person under sanctions'
```
Comment (criticus) [inline business_register/models/sanction_models.py:99]: typo

File: business_register/models/sanction_models.py [hunk 5]
```diff
@@ -28,84 +28,97 @@ class Meta:
...
+    )
+    last_name = models.CharField(
+        _('surname'),
+        max_length=30,
+        help_text='last name of of the person under sanctions'
+    )
+    fullname = models.CharField(
+        _("full name"),
```
Comment (criticus) [inline business_register/models/sanction_models.py:107]: please use single quotes here

File: business_register/models/sanction_models.py [hunk 6]
```diff
@@ -121,30 +134,15 @@ class Sanction(DataOceanModel):
         null=True,
         blank=True,
         default=None,
-        help_text=_('place of registration or residence of the person or company under sanctions')
+        help_text=_('place of registration or residence of the person under sanctions')
```
Comment (criticus) [inline business_register/models/sanction_models.py:137]: should this be a place of birth?

File: business_register/models/sanction_models.py [hunk 7]
```diff
@@ -169,38 +167,92 @@ class Sanction(DataOceanModel):
...
+        verbose_name = _('Sanction against person')
+        verbose_name_plural = _('Sanctions against persons')
+        ordering = ['start_date']
+
+
+class CompanySanction(Sanction):
+    is_foreign = models.BooleanField(
+        _('is foreign'),
+        db_index=True,
+        help_text=_('is foreign or Ukrainian')
```
Comment (criticus) [inline business_register/models/sanction_models.py:183]: consider adding a default value

File: business_register/models/sanction_models.py [hunk 8]
```diff
@@ -169,38 +167,92 @@ class Sanction(DataOceanModel):
...
-    # Maybe, we should established a different model later
-    imposed_by = models.CharField(
-        _('imposed by'),
-        max_length=50,
-        help_text=_('institution that imposed sanctions')
+    company = models.ForeignKey(
+        Company,
+        on_delete=models.CASCADE,
+        related_name='official_sanctions',
+        null=True,
```
Comment (criticus) [inline business_register/models/sanction_models.py:189]: could this be null?

File: business_register/models/sanction_models.py (hunk 1 above)
Comment (IegorSoboliev) [reply business_register/models/sanction_models.py:74]: have no idea what should be the value)

File: business_register/models/sanction_models.py (hunk 2 above)
Comment (IegorSoboliev) [reply business_register/models/sanction_models.py:76]: Yes, through related names

File: business_register/models/sanction_models.py (hunk 3 above)
Comment (IegorSoboliev) [reply business_register/models/sanction_models.py:94]: fixed

File: business_register/models/sanction_models.py (hunk 4 above)
Comment (IegorSoboliev) [reply business_register/models/sanction_models.py:99]: fixed

File: business_register/models/sanction_models.py (hunk 5 above)
Comment (IegorSoboliev) [reply business_register/models/sanction_models.py:107]: done

File: business_register/models/sanction_models.py (hunk 6 above)
Comment (IegorSoboliev) [reply business_register/models/sanction_models.py:137]: no, usually this is a place of registration from the passport or residence from open sources)

File: business_register/models/sanction_models.py (hunk 7 above)
Comment (IegorSoboliev) [reply business_register/models/sanction_models.py:183]: we really don`t know the proportion yet

File: business_register/models/sanction_models.py [hunk 9]
```diff
@@ -121,30 +134,14 @@ class Sanction(DataOceanModel):
...
-        _('registration number'),
-        max_length=15,
-        null=True,
-        blank=True,
-        default=None,
-        help_text=_('number of registration of the company under sanctions')
+        help_text=_('place of registration or residence of the person under sanctions')
     )
-    country_of_origin = models.ForeignKey(
+    country_of_citizenship = models.ManyToManyField(
```
Comment (roman-tiukh) [inline business_register/models/sanction_models.py:139]: ManyToManyField? One person can have many `country_of_citizenship`?

Comment (IegorSoboliev) [reply business_register/models/sanction_models.py:139]: Yes. We have discussed that)

Comment (IegorSoboliev) [reply business_register/models/sanction_models.py:139]: from source:
Альперін Вадим Олександрович (Альперин Вадим Александрович), 27.01.1972 р.н., громадянин України, зареєстрований за адресою: м. Одеса, просп. Добровольського, 134, кв. 33, паспорти громадянина України для виїзду за кордон: ES612183, FM192111 (ALPERIN VADYM), паспорт громадянина Держави Ізраїль 21793096 (ALPERIN VADIM), ІНПП 2632420633

**Review by roman-tiukh — APPROVED** [review]

File: business_register/models/sanction_models.py (hunk 1 above)
Comment (criticus) [reply business_register/models/sanction_models.py:74]: False :)

File: business_register/models/sanction_models.py (hunk 7 above)
Comment (criticus) [reply business_register/models/sanction_models.py:182]: actual proportion doesn't matter, it's just a good idea to have a default value for boolean field, especially if we're going to enter data manually via the admin side

**Review by criticus — APPROVED** [review]
looks good with minor comments

**Comment by roman-tiukh** [comment]
Adding `blank=True, default=False/True` its a good practice for BooleanField

File: business_register/models/sanction_models.py (hunk 7 above)
Comment (IegorSoboliev) [reply business_register/models/sanction_models.py:199]: added

File: data_converter/urls.py [hunk 10]
```diff
@@ -74,7 +76,9 @@
 router.register(r'historical-termination-started', HistoricalTerminationStartedView,
                 basename='historical_termination_started')
 router.register(r'pep', PepViewSet, basename='pep')
-router.register(r'sanction', SanctionViewSet, basename='sanction')
+router.register(r'country-sanction', CountrySanctionViewSet, basename='country_sanction')
+router.register(r'person-sanction', PersonSanctionViewSet, basename='person_sanction')
+router.register(r'company-sanction', CompanySanctionViewSet, basename='company_sanction')
```
Comment (roman-tiukh) [inline data_converter/urls.py:81]:
```suggestion
router.register('sanction/country', CountrySanctionViewSet, basename='country_sanction')
router.register('sanction/person', PersonSanctionViewSet, basename='person_sanction')
router.register('sanction/company', CompanySanctionViewSet, basename='company_sanction')
```

File: business_register/views/sanction_views.py [hunk 11]
```diff
@@ -3,31 +3,64 @@
...
+
+@method_decorator(name='retrieve', decorator=swagger_auto_schema(tags=['person-sanction']))
+@method_decorator(name='list', decorator=swagger_auto_schema(tags=['person-sanction']))
+class PersonSanctionViewSet(RegisterViewMixin,
+                            CachedViewSetMixin,
+                            viewsets.ReadOnlyModelViewSet):
+    permission_classes = [RegisterViewMixin.permission_classes[0]]
+    queryset = PersonSanction.objects.filter(is_confirmed=True)
+    serializer_class = PersonSanctionSerializer
+    filter_backends = (DODjangoFilterBackend, SearchFilter)
```
Comment (roman-tiukh) [inline business_register/views/sanction_views.py:37]: `DODjangoFilterBackend` is connected but `filterset_class` is not

File: business_register/views/sanction_views.py [hunk 12]
```diff
@@ -3,31 +3,64 @@
...
 from data_ocean.views import CachedViewSetMixin, RegisterViewMixin
 
 
-@method_decorator(name='retrieve', decorator=swagger_auto_schema(tags=['sanction']))
-@method_decorator(name='list', decorator=swagger_auto_schema(tags=['sanction']))
-class SanctionViewSet(RegisterViewMixin,
-                      CachedViewSetMixin,
-                      viewsets.ReadOnlyModelViewSet):
+@method_decorator(name='retrieve', decorator=swagger_auto_schema(tags=['country-sanction']))
+@method_decorator(name='list', decorator=swagger_auto_schema(tags=['country-sanction']))
```
Comment (roman-tiukh) [inline business_register/views/sanction_views.py:15]: `tags=['country-sanction']` maybe `sanction/country`? if its working

File: business_register/models/sanction_models.py [hunk 13]
```diff
@@ -28,84 +28,107 @@ class Meta:
 
 
 class Sanction(DataOceanModel):
```
Comment (roman-tiukh) [inline business_register/models/sanction_models.py:30]:
```suggestion
class BaseSanction(DataOceanModel):
```

File: business_register/models/sanction_models.py [hunk 14]
```diff
@@ -28,84 +28,107 @@ class Meta:
...
+
+
+class CountrySanction(Sanction):
+    country = models.ForeignKey(
+        Country,
         on_delete=models.CASCADE,
         related_name='official_sanctions',
         null=True,
         blank=True,
         default=None,
```
Comment (roman-tiukh) [inline business_register/models/sanction_models.py:59]: country can be blank?

File: business_register/models/sanction_models.py [hunk 15]
```diff
@@ -121,30 +144,13 @@ class Sanction(DataOceanModel):
...
+    countries_of_citizenship = models.ManyToManyField(
         Country,
-        on_delete=models.CASCADE,
-        null=True,
-        related_name='sanctions_applied_to_residents',
-        verbose_name=_('country of origin'),
-        help_text=_('country of citizenship/registration of the person or company under sanctions'))
+        related_name='sanctions_against_citizens',
+        verbose_name=_('country of citizenship'),
+        help_text=_('country of citizenship of the person under sanctions'))
```
Comment (roman-tiukh) [inline business_register/models/sanction_models.py:153]:
```suggestion
        help_text=_('country of citizenship of the person under sanctions')
    )
```

File: data_converter/urls.py (hunk 10 above)
Comment (roman-tiukh) [reply data_converter/urls.py:81]: need to remove all `r` string modificatiors, they redundant

File: business_register/models/sanction_models.py (hunk 1 above)
Comment (IegorSoboliev) [reply business_register/models/sanction_models.py:83]: done

File: business_register/views/sanction_views.py [hunk 16]
```diff
@@ -3,31 +3,64 @@
...
+        'countries_of_citizenship__name'
+    )
+
+
+@method_decorator(name='retrieve', decorator=swagger_auto_schema(tags=['company-sanction']))
+@method_decorator(name='list', decorator=swagger_auto_schema(tags=['company-sanction']))
+class CompanySanctionViewSet(RegisterViewMixin,
+                             CachedViewSetMixin,
+                             viewsets.ReadOnlyModelViewSet):
+    permission_classes = [RegisterViewMixin.permission_classes[0]]
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply business_register/views/sanction_views.py:55]: deleted

File: business_register/views/sanction_views.py [hunk 17]
```diff
@@ -3,31 +3,64 @@
...
+    queryset = CompanySanction.objects.filter(is_confirmed=True)
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply business_register/views/sanction_views.py:56]: made all()

File: business_register/models/sanction_models.py (hunk 13 above)
Comment (IegorSoboliev) [reply business_register/models/sanction_models.py:30]: done

**Review by criticus — APPROVED** [review]
looks good

File: business_register/admin.py [hunk 18]
```diff
@@ -93,9 +94,10 @@ def has_delete_permission(self, request, obj=None):
 class CompanySanctionAdmin(RegisterModelAdmin):
     list_display = (
         'name',
-        'number',
+        'taxpayer_number',
```
Comment (Hedgehog18) [inline business_register/admin.py:97]: The "number" field is the correct name, it means the registration number.
The "taxpayer number" field does not need to be a separate field. These are single data, they are rare (for example: "office location", "residence", etc.). As we discussed at the meeting - for such data it is better to make a field called "notes" or "additional data".

**Pull request merged**